lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -311,3 +312,83 @@ ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
   }
   assert(isValid());
 }
+
+/// Derive a concrete type of self and conformance from protocol conformance
+/// analysis If successful, initializes a valid ConcreteExistentialInfo.
+ConcreteExistentialInfo::ConcreteExistentialInfo(
+    ProtocolConformanceAnalysis *PCA, ClassHierarchyAnalysis *CHA,
+    SILBuilder &Builder, ProtocolDecl *Protocol, FullApplySite Apply,
+    SILValue &Arg, Operand &ArgOperand) {
+
+  // Determine the sole conforming type.
+  auto *NTD = PCA->findSoleConformingType(Protocol);
+  if (!NTD)
+    return;
+
+  // Sole conforming class should not be public or have derived classes.
+  ClassDecl *CD;
+  if ((CD = dyn_cast<ClassDecl>(NTD)) &&
+      (CD->getEffectiveAccess() == AccessLevel::Open ||
+       CHA->hasKnownDirectSubclasses(CD) ||
+       CHA->hasKnownIndirectSubclasses(CD))) {
```

Comment:
Out of curiosity, how can a class have indirect subclasses without having direct subclasses?

Reply:
My understanding was that they were disjoint sets. But looking at the code closely seems like Indirect Subclasses are derived from DirectSubClasses. So, check for hasKnownDirectSubClasses should suffice in this case.